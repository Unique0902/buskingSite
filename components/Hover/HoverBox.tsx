import React, {
  createContext,
  ReactNode,
  useContext,
  useEffect,
  useState,
} from 'react';

interface Props {
  children: ReactNode;
  hoverWaitSecond?: number;
}

type HoverBoxContextProps = {
  isShowElement: boolean;
};

const HoverBoxContext = createContext<HoverBoxContextProps>(
  {} as HoverBoxContextProps
);

const HoverBox = ({ children, hoverWaitSecond = 0.2 }: Props) => {
  const [isHovering, setIsHovering] = useState<boolean>(false);
  const [isShowElement, setIsShowElement] = useState<boolean>(false);
  useEffect(() => {
    !isHovering && setIsShowElement(false);
    const timer = isHovering
      ? setTimeout(() => setIsShowElement(true), hoverWaitSecond * 1000)
      : undefined;
    return () => clearTimeout(timer);
  }, [isHovering]); // hovered가 변할 때만 작동하도록
  return (
    <HoverBoxContext.Provider value={{ isShowElement }}>
      <div
        className='relative'
        onMouseOver={() => setIsHovering(true)}
        onMouseOut={() => setIsHovering(false)}
      >
        {children}
      </div>
    </HoverBoxContext.Provider>
  );
};

const OutElement = ({ children }: { children: ReactNode }) => {
  return <>{children}</>;
};

const InnerElement = ({ children }: { children: ReactNode }) => {
  const { isShowElement } = useContext(HoverBoxContext);
  return isShowElement ? <>{children}</> : <></>;
};

HoverBox.OutElement = OutElement;
HoverBox.InnerElement = InnerElement;

export default HoverBox;
