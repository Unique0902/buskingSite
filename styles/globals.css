@tailwind base;
@tailwind components;
@tailwind utilities;
/* TODO: 글꼴 불러오는 로딩 줄이거나 없애기 */
* {
  box-sizing: border-box;
  font-family: 'Pretendard';
}

@font-face {
  font-family: 'Pretendard';
  font-weight: 45 920;
  font-style: normal;
  font-display: swap;
  src: url('../assets/fonts/PretendardVariable.woff2')
    format('woff2-variations');
}

li {
  list-style: none;
}
body {
  margin: 0;
  @apply text-black dark:text-white dark:bg-slate-600;
}
p,
h1,
h2 {
  margin: 0;
}
ul {
  padding: 0;
}
