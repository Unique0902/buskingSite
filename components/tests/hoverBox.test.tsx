import { screen, render, waitFor } from '@testing-library/react';
import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import HoverBox from '../Hover/HoverBox';
import { expect, describe, it, jest } from '@jest/globals';

describe('arrangeMenu component test', () => {
  it('renders correctly', () => {
    const component = renderer.create(
      <HoverBox>
        <HoverBox.OutElement>
          <div>outer</div>
        </HoverBox.OutElement>
        <HoverBox.InnerElement>
          <div>inner</div>
        </HoverBox.InnerElement>
      </HoverBox>
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('hover outer box and show inner box correctly', async () => {
    render(
      <HoverBox hoverWaitSecond={0}>
        <HoverBox.OutElement>
          <div>outer</div>
        </HoverBox.OutElement>
        <HoverBox.InnerElement>
          <div>inner</div>
        </HoverBox.InnerElement>
      </HoverBox>
    );

    const outerBox = screen.getByText('outer');
    await userEvent.hover(outerBox);
    await waitFor(() => {
      expect(screen.getByText('inner')).toBeInTheDocument();
    });
  });

  it('hover outer box and show inner box correctly with time after 0.5sec', async () => {
    render(
      <HoverBox hoverWaitSecond={0.5}>
        <HoverBox.OutElement>
          <div>outer</div>
        </HoverBox.OutElement>
        <HoverBox.InnerElement>
          <div>inner</div>
        </HoverBox.InnerElement>
      </HoverBox>
    );

    const outerBox = screen.getByText('outer');
    await userEvent.hover(outerBox);
    await waitFor(() => {
      expect(screen.getByText('inner')).toBeInTheDocument();
    });
  });
});
