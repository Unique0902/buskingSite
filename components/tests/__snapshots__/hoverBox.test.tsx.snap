// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`arrangeMenu component test renders correctly 1`] = `
<div
  className="relative "
  onMouseOut={[Function]}
  onMouseOver={[Function]}
>
  <div>
    outer
  </div>
</div>
`;
